xapian-core/cluster/lcd_clusterer.cc

```diff
@@ -0,0 +1,164 @@
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include "xapian/error.h"
+
+#include "debuglog.h"
+
+#include <algorithm>
+#include <vector>
+
+using namespace Xapian;
+using namespace std;
+
+typedef set<pair<Point, double> > PSet;
+
+LCDClusterer::LCDClusterer(unsigned int k_)
+    : k(k_)
+{
+    LOGCALL_CTOR(API, "LCDClusterer", k_);
+    if (k_ == 0)
+	throw InvalidArgumentError("Number of required clusters should be "
+				   "greater than zero");
+}
+
+string
+LCDClusterer::get_description() const
+{
+    return "LCDClusterer()";
+}
+
+struct pcompare {
+    bool operator() (const pair<Point, double> &a,
+		     const pair<Point, double> &b) const {
+	return a.second > b.second;
+    }
+};
+
+struct dcompare {
+    bool operator() (const pair<PSet::iterator, double> &a,
+		     const pair<PSet::iterator, double> &b) const {
+	return a.second < b.second;
+    }
+};
+
+ClusterSet
+LCDClusterer::cluster(const MSet &mset)
+{
+    LOGCALL(API, ClusterSet, "LCDClusterer::cluster", mset);
+
+    doccount size = mset.size();
+    unsigned int k_ = k;
+    if (k_ >= size)
+	k_ = size;
+
+    // Store each document and its rel score from given mset
+    set<pair<Point, double>, pcompare> points;
+
+    // Initiliase points
+    TermListGroup tlg(mset);
+    for (MSetIterator it = mset.begin(); it != mset.end(); ++it)
+	points.emplace(Point(tlg, it.get_document()), it.get_weight());
+
+    // Container for holding the clusters
+    ClusterSet cset;
+
+    // Instantiate this for computing distance between points
+    CosineDistance distance;
+
+    // First cluster center
+    PSet::iterator cluster_center = points.begin();
+
+    for (unsigned int cnum = 1; cnum <= k_; ++cnum) {
+	// Container for new cluster
+	Cluster new_cluster;
+
+	// The original algorithm accepts a parameter 'k' which is the number
+	// of documents in each cluster, which can be hard to tune and
+	// especially when using this in conjunction with the diversification
+	// module. So, for now LCD clustering accepts 'k' as the number of
+	// clusters and divides the documents equally in the first k-1 clusters
+	// and the remaining in the last cluster. This needs to be tested on a
+	// dataset to see how well this works.
+	unsigned int num_points = size / k_;
+	if (cnum == k_)
+	    num_points = size - (k_ - 1) * (size / k_);
```

Comment:
Integer division truncates towards zero, so if `k_` is 5 and `num_points` is 24, then the first 4 clusters will be of size 4 and the last of size 8.

Having one cluster twice the size of the others seems unlikely to be helpful, and I wonder if we want to aim for a more even split such that the cluster sizes differ by at most one?

In this case 4 clusters of size 5 and one of size 4, but more generally `n` clusters of size `x` and `k_ - n` of size `x - 1`.